### PR TITLE
Uasr python scripts

### PR DESCRIPTION
**Pull request by DongjiGao**

### CODE REVIEW

File: espnet2/uasr/espnet_model.py [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+    from torch.cuda.amp import autocast
+else:
+    # Nothing to do if torch<1.6.0
+    @contextmanager
+    def autocast(enabled=True):
+        yield
+
+
+class ESPnetUASRModel(AbsESPnetModel):
+    """Unsupervised ASR model"""
```
Comment (ftshijt) [inline espnet2/uasr/espnet_model.py:43]: We should somehow acknowledge the original code source from fairseq

File: espnet2/tasks/uasr.py [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+            "--fairseq_checkpoint",
+            type=str,
+            help="Fairseq checkpoint to initialize model",
+        )
+
+        parser.add_argument(
+            "--ll_increasing_type",
+            type=str,
+            help="lattice loss increasing type",
+        )
```
[the remark this answers is not in the document]
Comment (DongjiGao) [reply espnet2/tasks/uasr.py:232]: fixed

File: espnet2/tasks/uasr.py [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+            for ctr in args.losses:
+                logging.info("initialize loss: {}".format(ctr["name"]))
+                if ctr["name"] == "gradient_penalty":
+                    loss = loss_choices.get_class(ctr["name"])(
+                        discriminator=discriminator, **ctr["conf"]
+                    )
+                elif ctr["name"] == "autoencoder_loss":
+                    loss = loss_choices.get_class(ctr["name"])(
+                        input_dim=vocab_size, **ctr["conf"]
+                    )
```
[the remark this answers is not in the document]
Comment (DongjiGao) [reply espnet2/tasks/uasr.py:372]: fixed

File: espnet2/uasr/espnet_model.py [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+from espnet2.asr.frontend.abs_frontend import AbsFrontend
+from espnet2.uasr.discriminator.abs_discriminator import AbsDiscriminator
+from espnet2.uasr.generator.abs_generator import AbsGenerator
+from espnet2.uasr.segmenter.abs_segmenter import AbsSegmenter
+
+from espnet2.uasr.loss.discriminator_loss import UASRDiscriminatorLoss
+from espnet2.uasr.loss.gradient_penalty import UASRGradientPenalty
+from espnet2.uasr.loss.phoneme_diversity_loss import UASRPhonemeDiversityLoss
+from espnet2.uasr.loss.pseudo_label_loss import UASRPseudoLabelLoss
+from espnet2.uasr.loss.smoothness_penalty import UASRSmoothnessPenalty
```
[the remark this answers is not in the document]
Comment (DongjiGao) [reply espnet2/uasr/espnet_model.py:27]: Do you mean from espnet2.uasr.loss.**abs_loss** import AbsUASRLoss

File: espnet2/uasr/espnet_model.py [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+                    batch_lm_log_prob += lm_log_prob
+                    batch_num_hyp_tokens += len(hyp_token_list)
+
+                    hyp_tokens_index = hyp_ids[hyp_ids >= 4]
+                    vocab_seen[hyp_tokens_index - 4] = True
+
+                ref_ids = ref_ids[ref_ids != self.pad]
+                ref_ids_list = ref_ids.tolist()
+                num_errors = editdistance.eval(hyp_ids_nosil_list, ref_ids_list)
+                batch_num_errors += num_errors
```
[the remark this answers is not in the document]
Comment (DongjiGao) [reply espnet2/uasr/espnet_model.py:239]: I feel like it would be better to keep the dev set to have text for consistency reasons (even some fake texts). I suggest adding an option for showing the PER since it is meaningless for fake texts.

File: espnet2/uasr/espnet_model.py [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+        # Collect total loss stats
+        stats["loss"] = loss.detach()
+        stats["generated_sample_prediction_loss"] = generated_sample_prediction_loss
+        stats["real_sample_prediction_loss"] = real_sample_prediction_loss
+        stats["gp"] = gp
+        stats["sp"] = sp
+        stats["pdl"] = pdl
+        stats["mmi"] = mmi
+        stats["ll"] = ll
+        stats["llwr"] = llwr
```
[the remark this answers is not in the document]
Comment (DongjiGao) [reply espnet2/uasr/espnet_model.py:374]: deleted

File: espnet2/uasr/espnet_model.py [hunk 7]
```diff
@@ -0,0 +1,476 @@
...
+class ESPnetUASRModel(AbsESPnetModel):
+    """Unsupervised ASR model"""
+
+    def __init__(
+        self,
+        frontend: Optional[AbsFrontend],
+        segmenter: Optional[AbsSegmenter],
+        generator: AbsGenerator,
+        discriminator: AbsDiscriminator,
+        losses: Dict[str, torch.nn.Module],
```
[the remark this answers is not in the document]
Comment (DongjiGao) [reply espnet2/uasr/espnet_model.py:51]: fixed

File: espnet2/uasr/espnet_model.py [hunk 8]
```diff
@@ -0,0 +1,476 @@
...
+
+    def get_ll_weight_ratio(self):
+        type = self.ll_increasing_type
+        if type == "constant":
+            llwr = 1
+        elif type == "linear":
+            llwr = self.number_epochs / self.max_epoch
+        if type == "quadratic":
+            llwr = (self.number_epochs / self.max_epoch) ** 2
+        return llwr
```
[the remark this answers is not in the document]
Comment (DongjiGao) [reply espnet2/uasr/espnet_model.py:476]: deleted

File: espnet2/uasr/espnet_model.py (hunk 1 above)
Comment (DongjiGao) [reply espnet2/uasr/espnet_model.py:43]: done

File: espnet2/uasr/espnet_model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ftshijt) [reply espnet2/uasr/espnet_model.py]: yeah
